core/comments/comment_view.ts

```diff
@@ -0,0 +1,686 @@
+
+  /** The text area where the user can type. */
+  private textArea: HTMLTextAreaElement;
+
+  /** The current size of the comment. */
+  private size: Size = new Size(120, 100);
+
+  /** Whether the comment is collapsed or not. */
+  private collapsed: boolean = false;
+
+  /** Whether the comment is editable or not. */
+  private editable: boolean = true;
+
+  /** The current location of the comment. */
+  private location: Coordinate = new Coordinate(0, 0);
+
+  /** The current text of the comment. Updates on  text area change. */
+  private text: string = '';
+
+  /** Listeners for changes to text. */
+  private textChangeListeners: Array<
+    (oldText: string, newText: string) => void
+  > = [];
+
+  /** Listeners for changes to size. */
+  private sizeChangeListeners: Array<(oldSize: Size, newSize: Size) => void> =
+    [];
+
+  /** Listeners for disposal. */
+  private disposeListeners: Array<() => void> = [];
+
+  /** Listeners for collapsing. */
+  private collapseChangeListeners: Array<(newCollapse: boolean) => void> = [];
+
+  /**
+   * Event data for the pointer up event on the resize handle. Used to
+   * unregister the listener.
+   */
+  private resizePointerUpListener: browserEvents.Data | null = null;
+
+  /**
+   * Event data for the pointer move event on the resize handle. Used to
+   * unregister the listener.
+   */
+  private resizePointerMoveListener: browserEvents.Data | null = null;
+
+  constructor(private readonly workspace: WorkspaceSvg) {
+    this.svgRoot = dom.createSvgElement(Svg.G, {
+      'class': 'blocklyComment blocklyEditable',
+    });
+
+    ({
+      topBar: this.topBar,
+      deleteIcon: this.deleteIcon,
+      foldoutIcon: this.foldoutIcon,
+      textPreview: this.textPreview,
+      textPreviewNode: this.textPreviewNode,
+    } = this.createTopBar(this.svgRoot, workspace));
+
+    ({foreignObject: this.foreignObject, textArea: this.textArea} =
+      this.createTextArea(this.svgRoot));
+
+    this.resizeHandle = this.createResizeHandle(this.svgRoot, workspace);
+
+    // TODO: Remove this comment before merging.
+    // I think we want comments to exist on the same layer as blocks.
+    workspace.getLayerManager()?.append(this, layers.BLOCK);
+
+    // Set size to the default size.
+    this.setSize(this.size);
+  }
+
+  /**
+   * Creates the top bar and the elements visually within it.
+   * Registers event listeners.
+   */
+  private createTopBar(
+    svgRoot: SVGGElement,
+    workspace: WorkspaceSvg,
+  ): {
+    topBar: SVGRectElement;
+    deleteIcon: SVGImageElement;
+    foldoutIcon: SVGImageElement;
+    textPreview: SVGTextElement;
+    textPreviewNode: Text;
+  } {
+    const topBar = dom.createSvgElement(
+      Svg.RECT,
+      {
+        'class': 'blocklyCommentTopbar',
+        'x': 0,
+        'y': 0,
+      },
+      svgRoot,
+    );
+    // TODO: Before merging, does this mean to override an individual image,
+    // folks need to replace the whole media folder?
```

Comment:
Do you have thoughts on this @rachel-fenichel ?

Reply:
I don't know. Is replacing those images an immediate need, or a theoretical one?

Reply:
It's a need for spork blocks. For MakeCode we can just include their image in core.

Reply:
@maribethb  as well.

Reply:
Yeah, they do, but tbh they probably should anyway because the default one links to one hosted on app engine. I don't see this as being a concern, it's how the rest of media assets work. If we wanted to refactor that in the future that's a different discussion and shouldn't be folded into this work imo

Reply:
Ok I'll dig into what MakeCode is currently doing and chat with them if it seems like it might be a problem. Since we've been holding v11 until we can have stuff ready for MakeCode, I want to make sure it actually is ready, and they're not going to have to wait for a v12.

Reply:
What kind of problems are you thinking about? The media directory is already configurable via options, and they can copy the assets they want to keep the same into their own hosting solution (which is what we currently recommend, though not that loudly). I feel like I'm missing something here if you're worried about something needing a v12.

Reply:
You're right I don't think this would need a v12. If anything I would just make it possible to configure the assets individually (which would be a feat not breaking). Based on some investigation it looks like they're already hosting everything themselves so it shouldn't be an issue. I just want to make sure we have everything pinned down so we can actually say MakeCode has been migrated.